File: src/background/creatureEquipTemplate.js
import { ipcMain } from "electron";

import {
  SEARCH_CREATURE_EQUIP_TEMPLATES,
  STORE_CREATURE_EQUIP_TEMPLATE,
  FIND_CREATURE_EQUIP_TEMPLATE,
  UPDATE_CREATURE_EQUIP_TEMPLATE,
  DESTROY_CREATURE_EQUIP_TEMPLATE,
  CREATE_CREATURE_EQUIP_TEMPLATE,
  COPY_CREATURE_EQUIP_TEMPLATE,
  GLOBAL_MESSAGE_BOX,
  GLOBAL_MESSAGE,
} from "../constants";

ipcMain.on(SEARCH_CREATURE_EQUIP_TEMPLATES, (event, payload) => {
  let queryBuilder = knex
    .select([
      "cet.*",
      "it1.name as name_1",
      "itl1.Name as localeName_1",
      "it1.Quality as Quality_1",
      "didi1.InventoryIcon_1 as Icon_1",
      "it2.name as name_2",
      "itl2.Name as localeName_2",
      "it2.Quality as Quality_2",
      "didi2.InventoryIcon_1 as Icon_2",
      "it3.name as name_3",
      "itl3.Name as localeName_3",
      "it3.Quality as Quality_3",
      "didi3.InventoryIcon_1 as Icon_3",
    ])
    .from("creature_equip_template as cet")
    .leftJoin("item_template as it1", "cet.ItemID1", "it1.entry")
    .leftJoin("item_template_locale as itl1", function () {
      this.on("cet.ItemID1", "=", "itl1.ID").andOn(
        "itl1.locale",
        "=",
        knex.raw("?", "zhCN")
      );
    })
    .leftJoin(
      "foxy.dbc_item_display_info as didi1",
      "it1.displayid",
      "didi1.ID"
    )
    .leftJoin("item_template as it2", "cet.ItemID2", "it2.entry")
    .leftJoin("item_template_locale as itl2", function () {
      this.on("cet.ItemID2", "=", "itl2.ID").andOn(
        "itl2.locale",
        "=",
        knex.raw("?", "zhCN")
      );
    })
    .leftJoin(
      "foxy.dbc_item_display_info as didi2",
      "it2.displayid",
      "didi2.ID"
    )
    .leftJoin("item_template as it3", "cet.ItemID3", "it3.entry")
    .leftJoin("item_template_locale as itl3", function () {
      this.on("cet.ItemID3", "=", "itl3.ID").andOn(
        "itl3.locale",
        "=",
        knex.raw("?", "zhCN")
      );
    })
    .leftJoin(
      "foxy.dbc_item_display_info as didi3",
      "it3.displayid",
      "didi3.ID"
    )
    .where(payload);

  queryBuilder
    .then((rows) => {
      event.reply(SEARCH_CREATURE_EQUIP_TEMPLATES, rows);
    })
    .catch((error) => {
      event.reply(`${SEARCH_CREATURE_EQUIP_TEMPLATES}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(STORE_CREATURE_EQUIP_TEMPLATE, (event, payload) => {
  let queryBuilder = knex.insert(payload).into("creature_equip_template");

  queryBuilder
    .then((rows) => {
      event.reply(STORE_CREATURE_EQUIP_TEMPLATE, rows);
    })
    .catch((error) => {
      event.reply(`${STORE_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(FIND_CREATURE_EQUIP_TEMPLATE, (event, payload) => {
  let queryBuilder = knex
    .select()
    .from("creature_equip_template")
    .where(payload);

  queryBuilder
    .then((rows) => {
      event.reply(FIND_CREATURE_EQUIP_TEMPLATE, rows.length > 0 ? rows[0] : {});
    })
    .catch((error) => {
      event.reply(`${FIND_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(UPDATE_CREATURE_EQUIP_TEMPLATE, (event, payload) => {
  let queryBuilder = knex
    .table("creature_equip_template")
    .where(payload.credential)
    .update(payload.creatureEquipTemplate);

  queryBuilder
    .then((rows) => {
      event.reply(UPDATE_CREATURE_EQUIP_TEMPLATE, rows);
    })
    .catch((error) => {
      event.reply(`${UPDATE_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(DESTROY_CREATURE_EQUIP_TEMPLATE, (event, payload) => {
  let queryBuilder = knex
    .table("creature_equip_template")
    .where(payload)
    .delete();

  queryBuilder
    .then((rows) => {
      event.reply(DESTROY_CREATURE_EQUIP_TEMPLATE, rows);
    })
    .catch((error) => {
      event.reply(`${DESTROY_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(CREATE_CREATURE_EQUIP_TEMPLATE, (event, payload) => {
  let queryBuilder = knex
    .select("ID")
    .from("creature_equip_template")
    .where(payload)
    .orderBy("ID", "desc");

  queryBuilder
    .then((rows) => {
      event.reply(CREATE_CREATURE_EQUIP_TEMPLATE, {
        CreatureID: payload.CreatureID,
        ID: rows.length > 0 ? rows[0].ID + 1 : 1,
      });
    })
    .catch((error) => {
      event.reply(`${CREATE_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    })
    .finally(() => {
      event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
    });
});

ipcMain.on(COPY_CREATURE_EQUIP_TEMPLATE, (event, payload) => {
  let id = undefined;
  let creatureEquipTemplate = undefined;

  let idQueryBuilder = knex
    .select("ID")
    .from("creature_equip_template")
    .where("CreatureID", payload.CreatureID)
    .orderBy("ID", "desc");
  let findCreatureEquipTemplateQueryBuilder = knex
    .select()
    .from("creature_equip_template")
    .where(payload);
  Promise.all([
    idQueryBuilder.then((rows) => {
      id = rows.length > 0 ? rows[0].ID : 1;
    }),
    findCreatureEquipTemplateQueryBuilder.then((rows) => {
      creatureEquipTemplate = rows.length > 0 ? rows[0] : {};
    }),
  ])
    .then(() => {
      creatureEquipTemplate.ID = id + 1;
      let queryBuilder = knex
        .insert(creatureEquipTemplate)
        .into("creature_equip_template");
      queryBuilder
        .then((rows) => {
          event.reply(COPY_CREATURE_EQUIP_TEMPLATE, rows);
        })
        .catch((error) => {
          event.reply(`${COPY_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
          event.reply(GLOBAL_MESSAGE_BOX, error);
        })
        .finally(() => {
          event.reply(GLOBAL_MESSAGE, queryBuilder.toString());
        });
    })
    .catch((error) => {
      event.reply(`${COPY_CREATURE_EQUIP_TEMPLATE}_REJECT`, error);
      event.reply(GLOBAL_MESSAGE_BOX, error);
    });
});

File: lib/model/creature_equip_template.dart
/// 生物装备模板
class CreatureEquipTemplate {
  final int creatureID;
  final int id;
  final int itemID1;
  final int itemID2;
  final int itemID3;
  final int verifiedBuild;

  // 关联字段（装备1信息）
  final String name1;
  final String localeName1;
  final int quality1;
  final String icon1;

  // 关联字段（装备2信息）
  final String name2;
  final String localeName2;
  final int quality2;
  final String icon2;

  // 关联字段（装备3信息）
  final String name3;
  final String localeName3;
  final int quality3;
  final String icon3;

  /// 显示名称
  String get displayName1 => localeName1.isNotEmpty ? localeName1 : name1;
  String get displayName2 => localeName2.isNotEmpty ? localeName2 : name2;
  String get displayName3 => localeName3.isNotEmpty ? localeName3 : name3;

  const CreatureEquipTemplate({
    this.creatureID = 0,
    this.id = 0,
    this.itemID1 = 0,
    this.itemID2 = 0,
    this.itemID3 = 0,
    this.verifiedBuild = 0,
    this.name1 = '',
    this.localeName1 = '',
    this.quality1 = 0,
    this.icon1 = '',
    this.name2 = '',
    this.localeName2 = '',
    this.quality2 = 0,
    this.icon2 = '',
    this.name3 = '',
    this.localeName3 = '',
    this.quality3 = 0,
    this.icon3 = '',
  });

  factory CreatureEquipTemplate.fromJson(Map<String, dynamic> json) {
    return CreatureEquipTemplate(
      creatureID: json['CreatureID'] ?? json['creatureID'] ?? 0,
      id: json['ID'] ?? json['id'] ?? 0,
      itemID1: json['ItemID1'] ?? json['itemID1'] ?? 0,
      itemID2: json['ItemID2'] ?? json['itemID2'] ?? 0,
      itemID3: json['ItemID3'] ?? json['itemID3'] ?? 0,
      verifiedBuild: json['VerifiedBuild'] ?? json['verifiedBuild'] ?? 0,
      name1: json['name_1'] ?? '',
      localeName1: json['localeName_1'] ?? '',
      quality1: json['Quality_1'] ?? 0,
      icon1: json['Icon_1'] ?? '',
      name2: json['name_2'] ?? '',
      localeName2: json['localeName_2'] ?? '',
      quality2: json['Quality_2'] ?? 0,
      icon2: json['Icon_2'] ?? '',
      name3: json['name_3'] ?? '',
      localeName3: json['localeName_3'] ?? '',
      quality3: json['Quality_3'] ?? 0,
      icon3: json['Icon_3'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'CreatureID': creatureID,
      'ID': id,
      'ItemID1': itemID1,
      'ItemID2': itemID2,
      'ItemID3': itemID3,
      'VerifiedBuild': verifiedBuild,
    };
  }
}
